x-pack/filebeat/input/netflow/decoder/v9/lru.go

```diff
@@ -0,0 +1,172 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package v9
+
+import (
+	"bytes"
+	"container/heap"
+	"sync"
+	"time"
+
+	"github.com/elastic/beats/v7/libbeat/common/atomic"
+)
+
+type eventWithMissingTemplate struct {
+	key       SessionKey
+	entryTime time.Time
+}
+
+type pendingEventsHeap []eventWithMissingTemplate
+
+func (h pendingEventsHeap) Len() int {
+	return len(h)
+}
+
+func (h pendingEventsHeap) Less(i, j int) bool {
+	return h[i].entryTime.Sub(h[j].entryTime) < 0
+}
+
+func (h pendingEventsHeap) Swap(i, j int) {
+	h[i], h[j] = h[j], h[i]
+}
+
+func (h *pendingEventsHeap) Push(x any) {
+	v, ok := x.(eventWithMissingTemplate)
+	if ok {
+		*h = append(*h, v)
+	}
+}
+
+func (h *pendingEventsHeap) Pop() any {
+	old := *h
+	n := len(old)
+	x := old[n-1]
+	*h = old[0 : n-1 : n-1]
+	return x
+}
+
+type pendingTemplatesCache struct {
+	mtx     sync.RWMutex
+	wg      sync.WaitGroup
+	isEmpty atomic.Bool
+	hp      pendingEventsHeap
+	started bool
+	events  map[SessionKey][]*bytes.Buffer
+}
+
+func newPendingTemplatesCache() *pendingTemplatesCache {
+	cache := &pendingTemplatesCache{
+		events: make(map[SessionKey][]*bytes.Buffer),
+		hp:     pendingEventsHeap{},
+	}
+	return cache
+}
+
+// GetAndRemove returns all events for a given session key and removes them from the cache
+func (h *pendingTemplatesCache) GetAndRemove(key SessionKey) []*bytes.Buffer {
+	if h == nil {
+		return nil
+	}
+
+	if h.isEmpty.Load() {
+		return nil
+	}
+
+	h.mtx.Lock()
+	defer h.mtx.Unlock()
+	events, ok := h.events[key]
+	if !ok {
+		return nil
+	}
+	delete(h.events, key)
+	h.isEmpty.Store(len(h.events) == 0)
+	return events
+}
+
+// Add adds an event to the pending templates cache
+func (h *pendingTemplatesCache) Add(key SessionKey, events *bytes.Buffer) {
+	if h == nil {
+		return
+	}
+
+	h.mtx.Lock()
+	defer h.mtx.Unlock()
+
+	h.events[key] = append(h.events[key], events)
+	h.hp.Push(eventWithMissingTemplate{key: key, entryTime: time.Now()})
+	h.isEmpty.Store(false)
+}
+
+// start starts the pending templates cache cleaner
+func (h *pendingTemplatesCache) start(done <-chan struct{}, cleanInterval time.Duration, removalThreshold time.Duration) {
+	if h == nil {
+		return
+	}
```

Comment:
nit: it's not typical in Go to add the nil checks in every receiver method unless there is a strong need to handle these specific cases. more often this a wrong use and should be exposed as a crash early.

Reply:
removed nil checks, feel to unresolve if you spot anything else